### PR TITLE
fix: explicit tile undelete when patching insights

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

In [this ticket ](https://posthoghelp.zendesk.com/agent/tickets/1290)a user's insight is in a very strange state where it reports being on dashboards despite the tile being deleted and won't let the user add or remove from dashboards

I can't replicate that locally but can intuit one possible error based on the code

## Changes

Fixes the possible error which should be fixed even if it isn't the bug here

## How did you test this code?

added a developer test and forced an insight into a broken state to test it locally

### CODE REVIEW

File: posthog/api/test/test_insight.py [hunk 1]
```diff
@@ -39,9 +39,16 @@ def setUp(self) -> None:
         self.dashboard_api = DashboardAPI(self.client, self.team, self.assertEqual)
 
     def test_get_insight_items(self) -> None:
-        filter_dict = {"events": [{"id": "$pageview"}], "properties": [{"key": "$browser", "value": "Mac OS X"}]}
+        filter_dict = {
```
Comment (pauldambra) [inline posthog/api/test/test_insight.py:42]: wow, black has gone to town on this file

File: posthog/api/test/test_insight.py [hunk 2]
```diff
@@ -455,10 +516,39 @@ def test_insight_items_on_a_dashboard_ignore_deleted_dashboards(self) -> None:
         new_dashboard_id, _ = self.dashboard_api.create_dashboard({})
         # accidentally include a deleted dashboard
         _, update_response = self.dashboard_api.update_insight(
-            insight_id, {"dashboards": [dashboard_id, deleted_dashboard_id, new_dashboard_id]}
+            insight_id,
+            {"dashboards": [dashboard_id, deleted_dashboard_id, new_dashboard_id]},
         )
         assert sorted(update_response["dashboards"]) == [dashboard_id, new_dashboard_id]
 
+    def test_insight_items_on_a_dashboard_ignore_deleted_dashboard_tiles(self) -> None:
```
Comment (pauldambra) [inline posthog/api/test/test_insight.py:524]: the change in this file only adding this test

File: posthog/api/test/dashboards/__snapshots__/test_dashboard.ambr [hunk 3]
```diff
@@ -268,6 +271,7 @@
          "posthog_team"."ingested_event",
          "posthog_team"."session_recording_opt_in",
          "posthog_team"."capture_console_log_opt_in",
+         "posthog_team"."capture_performance_opt_in",
```
Comment (pauldambra) [inline posthog/api/test/dashboards/__snapshots__/test_dashboard.ambr:274]: I'm not sure why this snapshot changed, but it isn't to do with this change so 🤷

Comment (neilkakkar) [reply posthog/api/test/dashboards/__snapshots__/test_dashboard.ambr:274]: I'm guessing the pr that added the migration didn't run these snapshots 🤔

File: posthog/api/test/test_insight.py (hunk 2 above)
Comment (neilkakkar) [reply posthog/api/test/test_insight.py:524]: what black version are you on?

File: posthog/api/insight.py [hunk 4]
```diff
@@ -310,7 +313,11 @@ def _update_insight_dashboards(self, dashboards: List[Dashboard], instance: Insi
         for dashboard in candidate_dashboards:
             if dashboard.team != instance.team:
                 raise serializers.ValidationError("Dashboard not found")
-            DashboardTile.objects.create(insight=instance, dashboard=dashboard)
+            tile, _ = DashboardTile.objects.get_or_create(insight=instance, dashboard=dashboard)
```
Comment (neilkakkar) [inline posthog/api/insight.py:316]: kind of confused, never touched this bit, what's the relation between a tile and an insight?

And what does a deleted tile mean?

Comment (pauldambra) [reply posthog/api/insight.py:316]: I think what has happened is that we've added a unique constraint to dashboard/tile in the DB when adding text cards. _After_ we wrote this code and the tests aren't forcing out where it is now fighting the DB.

Comment (neilkakkar) [reply posthog/api/insight.py:316]: Thanks, that's very useful to know!

**Review by neilkakkar — APPROVED** [review]

**Pull request merged**